xlive/Blam/Engine/input/input_abstraction.cpp

```diff
@@ -1,7 +1,442 @@
+}
+
+void __cdecl input_abstraction_get_player_look_angular_velocity_for_mouse(e_controller_index controller_index, real_euler_angles2d* angular_velocity)
+{
+	INVOKE(0x61CD3, 0x0, input_abstraction_get_player_look_angular_velocity_for_mouse, controller_index, angular_velocity);
+}
+
+bool __cdecl input_abstraction_controller_button_test(e_controller_index controller_index, e_button_functions button_index)
+{
+	return INVOKE(0x61C5B, 0x0, input_abstraction_controller_button_test, controller_index, button_index);
+}
+
+e_button_functions __cdecl input_abstraction_get_primary_fire_button(datum unit)
+{
+	return INVOKE(0x5E2B6, 0x0, input_abstraction_get_primary_fire_button, unit);
+}
+
+e_button_functions __cdecl input_abstraction_get_secondary_fire_button(datum unit)
+{
+	return INVOKE(0x5E2ED, 0x0, input_abstraction_get_secondary_fire_button, unit);
+}
+
 
 bool __cdecl input_abstraction_get_key_state(int16 key)
 {
 	return INVOKE(0x2EF86, 0x0, input_abstraction_get_key_state, key);
 }
+
+uint32 s_input_abstraction_globals_sub_45E501(e_button_functions button, void* a3)
+{
+	return INVOKE_TYPE(0x5E501, 0x0, uint32(__thiscall*)(s_input_abstraction_globals*, e_button_functions, void*), input_abstraction_globals, button, a3);
+}
+
+uint32 input_abstraction_get_stick_type_for_function(e_button_functions function)
+{
+	// Todo : define enum
+	// -1 = bad type
+	// 0  = left_stick,
+	// 1  = right_stick,
+	uint32 button = s_input_abstraction_globals_sub_45E501(function, NULL);
+	if (button > _gamepad_analog_left_stick_right)
+		return button > _gamepad_analog_right_stick_right ? NONE : 1;
+	return 0;
+}
+
+
+void input_abstraction_update_default_throttle(point2d* thumb, real_euler_angles2d* stick)
+{
+	real32 pointy = (real32)thumb->y;
+	real32 pointx = (real32)thumb->x;
+	real_angle angle = atan2(pointy, pointx);
+
+	real32 magnitude = MAX(fabs(sin(angle)), fabs(cos(angle)));
+	real32 inverse_magnitude = 1.0 / magnitude;
+
+	real32 normalized_yaw = (real32)(pointx * inverse_magnitude) * 0.000030518509f;
+	real32 normalized_pitch = (real32)(pointy * inverse_magnitude) * 0.000030518509f;
+
+
+	stick->yaw = PIN(normalized_yaw, -1.0f, 1.0f);
+	stick->pitch = PIN(normalized_pitch, -1.0f, 1.0f);
+
+}
+void input_abstraction_post_update_throttle(real_euler_angles2d* stick, real_angle angle, bool right_stick)
+{
+	real_angle flt_7B9F78[] =
+	{
+		0.78539819f,
+		2.3561945f,
+		-0.78539819f,
+		-2.3561945f
+	};
+
+	uint8 index;
+	if (stick->pitch >= 0.0f)
+		index = 0 + (stick->yaw < 0.0f);
+	else
+		index = 2 + (stick->yaw < 0.0f);
+
+	real_vector2d vec = { stick->yaw,stick->pitch };
+	real32 magnitude = square_root(dot_product2d(&vec, &vec));
+
+	real32 delta = fabs(angle - flt_7B9F78[index]);
+	real_angle min_delta = right_stick ? 0.17453292f : 0.61086524f;
+
+	if (delta >= min_delta)
+	{
+
+		float sign = 0.0f;
+		if (fabs(stick->yaw) <= fabs(stick->pitch))
+		{
+			if (stick->pitch >= 0.0f)
```

Comment:
can use tertiary operators for these statements as well as the ones below

Reply:
used ternary for calculating sign